[git]Revision e9638fa cannot be built with gcc 11.x: core_prefetch_buf.h:59:9: error: ‘memmove’ was not declared in this scope

**Describe the bug**
Hello.

I tried to build git at revision e9638fa. It stops with this error:

```
In file included from core_prefetch.cpp:135:
core_prefetch_buf.h: In function ‘void prefetch_lazyflush(Bitu)’:
core_prefetch_buf.h:59:9: error: ‘memmove’ was not declared in this scope
   59 |         memmove(prefetch_buffer,prefetch_buffer+prefetch_unit,pq_limit-prefetch_unit);
      |         ^~~~~~~
core_prefetch_buf.h:1:1: note: ‘memmove’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
  +++ |+#include <cstring>
    1 | 
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-int-to-void-pointer-cast’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-int-to-void-pointer-cast’ may have been intended to silence earlier diagnostics
make[4]: *** [Makefile:441: core_prefetch.o] Error 1
make[4]: *** Waiting for unfinished jobs....
```

Last buildable revision I got: c7a970a43

**To Reproduce**
Steps to reproduce the behavior:
1. Grab latest git revision
2. Use these configure options: --enable-core-inline --enable-debug --enable-avcodec --prefix=/usr --enable-sdl2
3. Launch build and wait
4. See error

**Expected behavior**
Build process going on.

**Environment (please complete the following information):**
 - Archlinux
 - DOSBox-X  commit SHA1: e9638fa
 - gcc version 11.1.0

It also breaks build process with clang 12.0.1:

```
In file included from core_prefetch.cpp:135:
./core_prefetch_buf.h:59:9: error: use of undeclared identifier 'memmove'; did you mean 'wmemmove'?
        memmove(prefetch_buffer,prefetch_buffer+prefetch_unit,pq_limit-prefetch_unit);
        ^~~~~~~
        wmemmove
/usr/include/bits/wchar2.h:68:8: note: 'wmemmove' declared here
__NTH (wmemmove (wchar_t *__s1, const wchar_t *__s2, size_t __n))
       ^
In file included from core_prefetch.cpp:135:
./core_prefetch_buf.h:59:17: error: cannot initialize a parameter of type 'wchar_t *' with an lvalue of type 'uint8_t [32]'
        memmove(prefetch_buffer,prefetch_buffer+prefetch_unit,pq_limit-prefetch_unit);
                ^~~~~~~~~~~~~~~
/usr/include/bits/wchar2.h:68:27: note: passing argument to parameter '__s1' here
__NTH (wmemmove (wchar_t *__s1, const wchar_t *__s2, size_t __n))
                          ^
2 errors generated.
make[4]: *** [Makefile:441: core_prefetch.o] Error 1
```

@Allofich: do you have any idea on how to fix this really annoying bug?

Thanks.

> Actually I tried the GitHub Action again now and it failed with the same error you had. I think that before I must have accidentally only tested it on the master branch and not on my changes. Well I'm actually happy about that because being able to reproduce the failure means I should be able to test fixing it myself. Sorry for the trouble.

I tried grabbing a clean copy of the latest git commit. 

I added back `#include <string.h>` to the top of src/cpu/core_prefetch.cpp.

Ran ./autogen.sh, and configure line.

Launch make -j1 and here is the result, this time breaking in core_dyn_x86.cpp:147

```
In file included from core_dyn_x86.cpp:147:
core_dyn_x86/cache.h: In member function 'void CodePageHandler::SetupAt(Bitu, PageHandler*)':
core_dyn_x86/cache.h:82:17: error: 'memset' was not declared in this scope
   82 |                 memset(&hash_map,0,sizeof(hash_map));
      |                 ^~~~~~
core_dyn_x86/cache.h:20:1: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
   19 | #include <assert.h>
  +++ |+#include <cstring>
   20 | 
core_dyn_x86/cache.h: In member function 'virtual void CodePageHandler::writeb(PhysPt, uint8_t)':
core_dyn_x86/cache.h:126:25: error: 'memset' was not declared in this scope
  126 |                         memset(invalidation_map,0,4096);
      |                         ^~~~~~
core_dyn_x86/cache.h:126:25: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
core_dyn_x86/cache.h: In member function 'virtual void CodePageHandler::writew(PhysPt, uint16_t)':
core_dyn_x86/cache.h:146:25: error: 'memset' was not declared in this scope
  146 |                         memset(invalidation_map,0,4096);
      |                         ^~~~~~
core_dyn_x86/cache.h:146:25: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
core_dyn_x86/cache.h: In member function 'virtual void CodePageHandler::writed(PhysPt, uint32_t)':
core_dyn_x86/cache.h:166:25: error: 'memset' was not declared in this scope
  166 |                         memset(invalidation_map,0,4096);
      |                         ^~~~~~
core_dyn_x86/cache.h:166:25: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
core_dyn_x86/cache.h: In member function 'virtual bool CodePageHandler::writeb_checked(PhysPt, uint8_t)':
core_dyn_x86/cache.h:186:33: error: 'memset' was not declared in this scope
  186 |                                 memset(invalidation_map,0,4096);
      |                                 ^~~~~~
core_dyn_x86/cache.h:186:33: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
core_dyn_x86/cache.h: In member function 'virtual bool CodePageHandler::writew_checked(PhysPt, uint16_t)':
core_dyn_x86/cache.h:212:33: error: 'memset' was not declared in this scope
  212 |                                 memset(invalidation_map,0,4096);
      |                                 ^~~~~~
core_dyn_x86/cache.h:212:33: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
core_dyn_x86/cache.h: In member function 'virtual bool CodePageHandler::writed_checked(PhysPt, uint32_t)':
core_dyn_x86/cache.h:238:33: error: 'memset' was not declared in this scope
  238 |                                 memset(invalidation_map,0,4096);
      |                                 ^~~~~~
core_dyn_x86/cache.h:238:33: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
In file included from core_dyn_x86/cache.h:507,
                 from core_dyn_x86.cpp:147:
../../src/cpu/dynamic_alloc_common.h: In function 'void cache_dynamic_common_alloc(Bitu)':
../../src/cpu/dynamic_alloc_common.h:187:107: error: 'strerror' was not declared in this scope; did you mean 'stderr'?
  187 | e permission on the code cache has failed (malloc)! err=%s",strerror(errno));
      |                                                             ^~~~~~~~
      |                                                             stderr
In file included from core_dyn_x86.cpp:147:
core_dyn_x86/cache.h: In function 'void cache_init(bool)':
core_dyn_x86/cache.h:526:25: error: 'memset' was not declared in this scope
  526 |                         memset(cache_blocks,0,sizeof(CacheBlock)*CACHE_BLOCKS);
      |                         ^~~~~~
core_dyn_x86/cache.h:526:25: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
core_dyn_x86/cache.h: In function 'void cache_reset()':
core_dyn_x86/cache.h:607:17: error: 'memset' was not declared in this scope
  607 |                 memset(cache_blocks,0,sizeof(CacheBlock)*CACHE_BLOCKS);
      |                 ^~~~~~
core_dyn_x86/cache.h:607:17: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
In file included from core_dyn_x86.cpp:232:
core_dyn_x86/risc_x64.h: In function 'void gen_call_function(void*, const char*, ...)':
core_dyn_x86/risc_x64.h:1070:9: error: 'va_start' was not declared in this scope
 1070 |         va_start(params,ops);
      |         ^~~~~~~~
core_dyn_x86/risc_x64.h:1078:96: error: expected primary-expression before ')' token
 1078 |      if (*ops++!='p') gen_load_imm(gen->index,va_arg(params,uint32_t));
      |                                                                     ^

core_dyn_x86/risc_x64.h:1078:74: error: 'va_arg' was not declared in this scope
 1078 |                    if (*ops++!='p') gen_load_imm(gen->index,va_arg(params,uint32_t));
      |                                                             ^~~~~~

core_dyn_x86/risc_x64.h:1079:80: error: expected primary-expression before ')' token
 1079 |                      else gen_load_imm(gen->index,va_arg(params,Bitu));
      |                                                                     ^

core_dyn_x86/risc_x64.h:1079:62: error: 'va_arg' was not declared in this scope
 1079 |                                else gen_load_imm(gen->index,va_arg(params,Bitu));
      |                                                             ^~~~~~

core_dyn_x86/risc_x64.h:1082:89: error: expected primary-expression before '*' token
 1082 |     gen_load_arg_reg((int)paramcount++, va_arg(params,DynReg*), ops++);
      |                                                             ^

core_dyn_x86/risc_x64.h:1082:90: error: expected primary-expression before ')' token
 1082 |     gen_load_arg_reg((int)paramcount++, va_arg(params,DynReg*), ops++);
      |                                                              ^

core_dyn_x86/risc_x64.h:1082:69: error: 'va_arg' was not declared in this scope
 1082 |                         gen_load_arg_reg((int)paramcount++, va_arg(params,DynReg*), ops++);
      |                                                             ^~~~~~

core_dyn_x86/risc_x64.h:1085:62: error: expected primary-expression before '*' token
 1085 |                                 dynret = va_arg(params,DynReg*);
      |                                                              ^
core_dyn_x86/risc_x64.h:1085:63: error: expected primary-expression before ')' token
 1085 |                                 dynret = va_arg(params,DynReg*);
      |                                                               ^
core_dyn_x86/risc_x64.h:1100:9: error: 'va_end' was not declared in this scope; did you mean 'var_read'?
 1100 |         va_end(params);
      |         ^~~~~~
      |         var_read
core_dyn_x86/risc_x64.h: In function 'void gen_fill_branch(uint8_t*, uint8_t*)':
core_dyn_x86/risc_x64.h:1163:36: warning: format '%d' expects argument of type 'int', but argument 2 has type 'Bits' {aka 'long int'} [-Wformat=]
 1163 |                 LOG_MSG("Big jump %d",len);
      |                                   ~^  ~~~
      |                                    |  |
      |                                    |  Bits {aka long int}
      |                                    int
      |                                   %ld
core_dyn_x86/risc_x64.h: In function 'void gen_fill_short_jump(uint8_t*, uint8_t*)':
core_dyn_x86/risc_x64.h:1199:36: warning: format '%d' expects argument of type 'int', but argument 2 has type 'Bits' {aka 'long int'} [-Wformat=]
 1199 |                 LOG_MSG("Big jump %d",len);
      |                                   ~^  ~~~
      |                                    |  |
      |                                    |  Bits {aka long int}
      |                                    int
      |                                   %ld
In file included from ../../include/logging.h:22,
                 from ../../include/fpu.h:22,
                 from core_dyn_x86.cpp:39:
core_dyn_x86/decoder.h: In function 'bool MakeCodePage(Bitu, CodePageHandler*&)':
core_dyn_x86/decoder.h:80:44: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
   80 |         if (GCC_UNLIKELY(mem_readb_checked((const PhysPt)lin_addr,&rdval))) return true;
      |                                            ^~~~~~~~~~~~~~~~~~~~~~
../../config.h:349:43: note: in definition of macro 'GCC_UNLIKELY'
  349 | #define GCC_UNLIKELY(x) __builtin_expect((x),0)
      |                                           ^
In file included from core_dyn_x86.cpp:290:
core_dyn_x86/decoder.h:81:51: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
   81 |         PageHandler * handler=get_tlb_readhandler((const PhysPt)lin_addr);
      |                                                   ^~~~~~~~~~~~~~~~~~~~~~
core_dyn_x86/decoder.h:87:45: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
   87 |                 handler=get_tlb_readhandler((const PhysPt)lin_addr);
      |                                             ^~~~~~~~~~~~~~~~~~~~~~
core_dyn_x86/decoder.h:91:53: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
   91 |                         handler=get_tlb_readhandler((const PhysPt)lin_addr);
      |                                                     ^~~~~~~~~~~~~~~~~~~~~~
core_dyn_x86/decoder.h:97:61: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
   97 |                                 handler=get_tlb_readhandler((const PhysPt)lin_addr);
      |                                                             ^~~~~~~~~~~~~~~~~~~~~~
core_dyn_x86/decoder.h: In function 'uint8_t decode_fetchb()':
core_dyn_x86/decoder.h:146:35: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
  146 |                         mem_readb((const PhysPt)fetchaddr);
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~
core_dyn_x86/decoder.h: In function 'void decode_increase_wmapmask(Bitu)':
core_dyn_x86/decoder.h:210:17: error: 'memset' was not declared in this scope
  210 |                 memset(activecb->cache.wmapmask,0,START_WMMEM);
      |                 ^~~~~~
core_dyn_x86/decoder.h:20:1: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
   19 | #include "inout.h"
  +++ |+#include <cstring>
   20 | 
core_dyn_x86/decoder.h:220:25: error: 'memset' was not declared in this scope
  220 |                         memset(tempmem,0,newmasklen);
      |                         ^~~~~~
core_dyn_x86/decoder.h:220:25: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
core_dyn_x86/decoder.h:221:25: error: 'memcpy' was not declared in this scope
  221 |                         memcpy(tempmem,activecb->cache.wmapmask,activecb->cache.masklen);
      |                         ^~~~~~
core_dyn_x86/decoder.h:221:25: note: 'memcpy' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
core_dyn_x86/decoder.h: In function 'void dyn_set_eip_end(DynReg*)':
core_dyn_x86/decoder.h:338:45: warning: unused parameter 'endreg' [-Wunused-parameter]
  338 | static INLINE void dyn_set_eip_end(DynReg * endreg) {
      |                                    ~~~~~~~~~^~~~~~
In file included from core_dyn_x86.cpp:290:
core_dyn_x86/decoder.h: In function 'CacheBlock* CreateCacheBlock(CodePageHandler*, PhysPt, Bitu)':
core_dyn_x86/decoder.h:2489:14: warning: unused variable 'cycles' [-Wunused-variable]
 2489 |         Bitu cycles=0;
      |              ^~~~~~
core_dyn_x86.cpp: In function 'void CPU_Core_Dyn_X86_Init()':
core_dyn_x86.cpp:535:9: error: 'memset' was not declared in this scope
  535 |         memset(&dyn_dh_fpu.state, 0, sizeof(dyn_dh_fpu.state));
      |         ^~~~~~
core_dyn_x86.cpp:291:1: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
  290 | #include "core_dyn_x86/decoder.h"
  +++ |+#include <cstring>
  291 | 
In file included from core_dyn_x86/decoder.h:2476,
                 from core_dyn_x86.cpp:290:
core_dyn_x86/dyn_mmx.h: At global scope:
core_dyn_x86/dyn_mmx.h:57:13: warning: 'void gen_mmx_op(Bitu, Bitu, Bitu, PhysPt)' defined but not used [-Wunused-function]
   57 | static void gen_mmx_op(Bitu op, Bitu rm, Bitu imm = 0, PhysPt eaa = 0) {
      |             ^~~~~~~~~~
In file included from core_dyn_x86.cpp:290:
core_dyn_x86/decoder.h:2366:13: warning: 'void dyn_interrupt(Bitu)' defined but not used [-Wunused-function]
 2366 | static void dyn_interrupt(Bitu num) {
      |             ^~~~~~~~~~~~~
core_dyn_x86/decoder.h:2181:13: warning: 'void dyn_normal_exit(BlockReturn)' defined but not used [-Wunused-function]
 2181 | static void dyn_normal_exit(BlockReturn code) {
      |             ^~~~~~~~~~~~~~~
In file included from core_dyn_x86.cpp:232:
core_dyn_x86/risc_x64.h:485:13: warning: 'void gen_clearzeroflag()' defined but not used [-Wunused-function]
  485 | static void gen_clearzeroflag(void) {
      |             ^~~~~~~~~~~~~~~~~
core_dyn_x86/risc_x64.h:480:13: warning: 'void gen_setzeroflag()' defined but not used [-Wunused-function]
  480 | static void gen_setzeroflag(void) {
      |             ^~~~~~~~~~~~~~~
cc1plus: note: unrecognized command-line option '-Wno-int-to-void-pointer-cast' may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option '-Wno-int-to-void-pointer-cast' may have been intended to silence earlier diagnostics
make[4]: *** [Makefile:441: core_dyn_x86.o] Error 1
```

Looks like the fix will be less simpler than you think first. Smells like you'll have to reverse this commit and try to apply it after 0.83.17 release.

--- a/src/cpu/core_dyn_x86.cpp
+++ b/src/cpu/core_dyn_x86.cpp
@@ -16,6 +16,9 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include <stdarg.h>
+#include <string.h>
+
 #include "config.h"
 
 #if (C_DYNAMIC_X86)

--- a/src/cpu/core_dynrec.cpp
+++ b/src/cpu/core_dynrec.cpp
@@ -19,6 +19,8 @@
 #include "config.h"
 
 #if (C_DYNREC)
+#include <string.h>
+
 #if defined (WIN32)
 #include <windows.h>
 #include <winbase.h>

--- a/src/cpu/core_normal_286.cpp
+++ b/src/cpu/core_normal_286.cpp
@@ -16,6 +16,8 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include <stdio.h>
+
 #include "cpu.h"
 #include "lazyflags.h"
 #include "callback.h"

--- a/src/cpu/core_prefetch.cpp
+++ b/src/cpu/core_prefetch.cpp
@@ -16,6 +16,9 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include <stdio.h>
+#include <string.h>
+
 #include "cpu.h"
 #include "lazyflags.h"
 #include "callback.h"

--- a/src/cpu/core_prefetch_286.cpp
+++ b/src/cpu/core_prefetch_286.cpp
@@ -16,6 +16,9 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include <stdio.h>
+#include <string.h>
+
 #include "cpu.h"
 #include "lazyflags.h"
 #include "callback.h"

--- a/src/cpu/core_prefetch_8086.cpp
+++ b/src/cpu/core_prefetch_8086.cpp
@@ -16,6 +16,9 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include <stdio.h>
+#include <string.h>
+
 #include "cpu.h"
 #include "lazyflags.h"
 #include "callback.h"

--- a/src/cpu/core_simple.cpp
+++ b/src/cpu/core_simple.cpp
@@ -16,6 +16,9 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include <stdio.h>
+#include <stdlib.h>
+
 #include "cpu.h"
 #include "lazyflags.h"
 #include "callback.h"

